From 652db04c7adcee9e7ce1ccfa8ac56daa1fb38ca2 Mon Sep 17 00:00:00 2001
From: Katsuya Matsubara <matsu@igel.co.jp>
Date: Wed, 20 Jun 2012 11:54:28 +0900
Subject: [PATCH 64/73] rcarvin: prevent duplicating UYVY entries in the VIDIOC_ENUM_FMT list

Duplicating UYVY entries in the VIDIOC_ENUM_FMT list could be induced
by appending the sensor native output format into the list.
Fortunately all the sensor output formats except RGB32 may be also
included in the VIN extra format list.
This removes the sensor format addition and appends RGB32 in case of
the RGB inputs.
---
 drivers/media/video/rcarvin.c |   31 ++++++++++++++-----------------
 1 files changed, 14 insertions(+), 17 deletions(-)

diff --git a/drivers/media/video/rcarvin.c b/drivers/media/video/rcarvin.c
index 4187969..cea3dd9 100644
--- a/drivers/media/video/rcarvin.c
+++ b/drivers/media/video/rcarvin.c
@@ -1193,19 +1193,26 @@ static int rcar_vin_get_formats(struct soc_camera_device *icd, unsigned int idx,
 	case V4L2_MBUS_FMT_RGB888_1X24_LE:
 	case V4L2_MBUS_FMT_RGB888_2X12_LE:
 	case V4L2_MBUS_FMT_RGB666_1X18_LE:
+		/* add the sensor native format 'RGB32' */
+		formats++;
+		if (xlate) {
+			xlate->host_fmt	= fmt;
+			xlate->code	= code;
+			xlate++;
+			dev_dbg(dev, "Providing format %s in pass-through mode\n",
+				xlate->host_fmt->name);
+		}
+		/* fall through */
 	case V4L2_MBUS_FMT_YUYV8_1X16_BE:
 	case V4L2_MBUS_FMT_YUYV8_2X8_BE:
 		if (cam->extra_fmt)
 			break;
 
 		/*
-		 * Our case is simple so far: for any of the above four camera
-		 * formats we add all our four synthesized NV* formats, so,
-		 * just marking the device with a single flag suffices. If
-		 * the format generation rules are more complex, you would have
-		 * to actually hang your already added / counted formats onto
-		 * the host_priv pointer and check whether the format you're
-		 * going to add now is already there.
+		 * Our case is simple so far: for the above two yuv mediabus formats
+		 * we add all our formats can be generated by VIN, since they must
+		 * include the sensor formats already. For RGB mediabus formats,
+		 * the sensor native format 'RGB32' should be appended into the list.
 		 */
 		cam->extra_fmt = rcar_vin_formats;
 
@@ -1224,16 +1231,6 @@ static int rcar_vin_get_formats(struct soc_camera_device *icd, unsigned int idx,
 			return 0;
 	}
 
-	/* Generic pass-through */
-	formats++;
-	if (xlate) {
-		xlate->host_fmt	= fmt;
-		xlate->code	= code;
-		xlate++;
-		dev_dbg(dev, "Providing format %s in pass-through mode\n",
-			xlate->host_fmt->name);
-	}
-
 	return formats;
 }
 
-- 
1.7.0.4

